sale_order_revision/models/sale_order.py

```diff
@@ -0,0 +1,119 @@
+    old_revision_ids = fields.One2many('sale.order',
+                                       'current_revision_id',
+                                       'Old revisions',
+                                       readonly=True,
+                                       context={'active_test': False})
+    revision_number = fields.Integer('Revision',
+                                     copy=False,
+                                     default=0)
+    unrevisioned_name = fields.Char('Order Reference',
+                                    copy=True,
+                                    readonly=True)
+    active = fields.Boolean('Active',
+                            default=True)
+
+    has_old_revisions = fields.Boolean('Has old revisions',
+                                       compute=_has_old_revisions)
+
+    _sql_constraints = [
+        ('revision_unique',
+         'unique(unrevisioned_name, revision_number, company_id)',
+         'Order Reference and revision must be unique per Company.'),
+    ]
+
+    @api.returns('self', lambda value: value.id)
+    @api.multi
+    def copy(self, default=None):
+        if default is None:
+            default = {}
+        if self.env.context.get('new_sale_revision'):
+
+            new_rev_number = self.revision_number + 1
+            default.update({
+                'revision_number': new_rev_number,
+                'unrevisioned_name': self.unrevisioned_name,
+                'name': '%s-%02d' % (self.unrevisioned_name, new_rev_number),
+                'old_revision_ids': [(4, self.id, False)],
+            })
+            new_revision = super(SaleOrder, self).copy(default)
+            self.old_revision_ids.write({
+                'current_revision_id': new_revision.id,
+            })
+            self.write({'active': False,
+                        'state': 'cancel',
+                        'current_revision_id': new_revision.id,
+                        })
+            return new_revision
+
+        else:
+            if default.get('name', '/') == '/':
+                seq = self.env['ir.sequence']
+                default['name'] = seq.next_by_code('sale.order') or '/'
+                default['revision_number'] = 0
+                default['unrevisioned_name'] = default['name']
+            return super(SaleOrder, self).copy(default=default)
+
+    @api.model
+    def create(self, values):
+        if 'unrevisioned_name' not in values:
+            if values.get('name', '/') == '/':
+                seq = self.env['ir.sequence']
+                values['name'] = seq.next_by_code('sale.order') or '/'
+            values['unrevisioned_name'] = values['name']
+        return super(SaleOrder, self).create(values)
+
+    @api.multi
+    def create_revision(self):
+
+        revision_ids = []
+        # Looping over sale order records
+        for sale_order_rec in self:
+            # Calling  Copy method
+            copied_sale_rec = sale_order_rec.with_context(
```

Comment:
As I said here upper, on each iteration you create a new environment at each call of 'with_context'. Prefer calling a new function like copy_with_new_revision or ...

Reply:
From that statement; we are creating revision by using copy method, and making some changes in a few fields..... and if the user duplicates the records from the action menu, it should also create a revision with different values in some other fields. 

I tried to use a separate method, but it is difficult to identify that revision creation is done by a method or via 'Duplicate' action.

Reply:
Ok, but call it on self not in the loop :
´´´
for sale_order_rec in self.with_context(...):
´´´

Reply:
Ok. Thank you for the reply. Meanwhile, I tried on your idea and somehow managed to do that without context, Hope that works well. Look in new commit